dotlottie-rs/src/lottie_renderer/thorvg.rs

```diff
@@ -13,77 +13,84 @@ mod tvg {
+impl From<ColorSpace> for tvg::Tvg_Colorspace {
+    fn from(color_space: ColorSpace) -> Self {
+        match color_space {
+            ColorSpace::ABGR8888 => tvg::Tvg_Colorspace_TVG_COLORSPACE_ABGR8888,
+            ColorSpace::ABGR8888S => tvg::Tvg_Colorspace_TVG_COLORSPACE_ABGR8888S,
+            ColorSpace::ARGB8888 => tvg::Tvg_Colorspace_TVG_COLORSPACE_ARGB8888,
+            ColorSpace::ARGB8888S => tvg::Tvg_Colorspace_TVG_COLORSPACE_ARGB8888S,
+        }
+    }
 }
 
 pub enum TvgEngine {
     TvgEngineSw,
     TvgEngineGl,
 }
 
-fn convert_tvg_result(result: tvg::Tvg_Result, function_name: &str) -> Result<(), TvgError> {
-    let func_name = function_name.to_string();
-
-    match result {
-        tvg::Tvg_Result_TVG_RESULT_SUCCESS => Ok(()),
-        tvg::Tvg_Result_TVG_RESULT_INVALID_ARGUMENT => Err(TvgError::InvalidArgument {
-            function_name: func_name,
-        }),
-        tvg::Tvg_Result_TVG_RESULT_INSUFFICIENT_CONDITION => Err(TvgError::InsufficientCondition {
-            function_name: func_name,
-        }),
-        tvg::Tvg_Result_TVG_RESULT_FAILED_ALLOCATION => Err(TvgError::FailedAllocation {
-            function_name: func_name,
-        }),
-        tvg::Tvg_Result_TVG_RESULT_MEMORY_CORRUPTION => Err(TvgError::MemoryCorruption {
-            function_name: func_name,
-        }),
-        tvg::Tvg_Result_TVG_RESULT_NOT_SUPPORTED => Err(TvgError::NotSupported {
-            function_name: func_name,
-        }),
-        tvg::Tvg_Result_TVG_RESULT_UNKNOWN => Err(TvgError::Unknown {
-            function_name: func_name,
-        }),
-        _ => Err(TvgError::Unknown {
-            function_name: func_name,
-        }),
+impl From<TvgEngine> for tvg::Tvg_Engine {
+    fn from(engine_method: TvgEngine) -> Self {
+        match engine_method {
+            TvgEngine::TvgEngineSw => tvg::Tvg_Engine_TVG_ENGINE_SW,
+            TvgEngine::TvgEngineGl => tvg::Tvg_Engine_TVG_ENGINE_GL,
+        }
     }
 }
 
+static RENDERERS_COUNT: AtomicUsize = AtomicUsize::new(0);
+
 pub struct TvgRenderer {
     raw_canvas: *mut tvg::Tvg_Canvas,
     engine_method: tvg::Tvg_Engine,
 }
 
 impl TvgRenderer {
     pub fn new(engine_method: TvgEngine, threads: u32) -> Self {
-        let engine = match engine_method {
-            TvgEngine::TvgEngineSw => tvg::Tvg_Engine_TVG_ENGINE_SW,
-            TvgEngine::TvgEngineGl => tvg::Tvg_Engine_TVG_ENGINE_GL,
-        };
-
-        let init_result = unsafe { tvg::tvg_engine_init(engine, threads) };
+        let engine = engine_method.into();
 
-        if init_result != tvg::Tvg_Result_TVG_RESULT_SUCCESS {
-            panic!("Failed to initialize ThorVG engine");
+        if RENDERERS_COUNT.load(std::sync::atomic::Ordering::SeqCst) == 0 {
+            unsafe { tvg::tvg_engine_init(engine, threads).into_result() }
+                .expect("Failed to initialize ThorVG engine");
         }
 
+        RENDERERS_COUNT.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
```

Comment:
There is a concurrency issue with this approach. Let's say we have two different treads calling `TvgRenderer::new`. Both of them happen to read the `RENDERERS_COUNT` variable *concurrently* and see that it's 0, which means that both of them will end up calling `tvg_engine_init`.

Instead, we need to make sure that both reading and incrementing the counter happens in one atomic op. This is what `fetch_add` achieves:

```suggestion
        if RENDERERS_COUNT.fetch_add(std::sync::atomic::Ordering::Relaxed) == 0 {
            unsafe { tvg::tvg_engine_init(engine, threads).into_result() }
                .expect("Failed to initialize ThorVG engine");
        }
```

Also, `SeqCst` is not needed here. We don't care if the instructions around the `RENDERERS_COUNT` fetch are re-ordered, we just need to make sure that we get a unique value, which is exactly what `Relaxed` offers us.

Reply:
There is also another issue here:

We have 2 threads. Both call `TvgRenderer::new`. Thread 1 calls `fetch_add`, sees 0 and then gets interrupted before it can call `tvg_engine_init` and put on pause. Thread 2 reaches the `fetch_add`, sees 1, thinks the engine is initialized and continues calling `tvg::*` functions that shouldn't be called until the engine is initialized.

To make this work, you need to lock onto the engine creation. I.e. all other threads need to wait until the engine was created which can't be done with *just* atomics. A `RwLock` or a `Mutex` could work instead.

Reply:
@dragostis Thanks so much for your review, it's been incredibly helpful!

I've updated the implementation here https://github.com/LottieFiles/dotlottie-rs/pull/265/commits/cd2f0f868263e6e70c3c0e3bd06c68526a88159d to replace the use of atomic operations for updating the RENDERER_COUNTER with a Mutex, ensuring that no race conditions can occur during engine initialization & termination. 

I've also added a test for that, but I'm not sure if it's good enough. I was thinking that i might need to stub `tvg_engine_init` and `tvg_engine_term` to verify that they are only called once across multiple threads, any ideas ?